ddtrace/contrib/graphql/patch.py

```diff
@@ -0,0 +1,150 @@
+import re
+from typing import Any
+from typing import Union
+
+import graphql
+from graphql.language.source import Source
+
+from ddtrace import Span
+from ddtrace import config
+from ddtrace.constants import ANALYTICS_SAMPLE_RATE_KEY
+from ddtrace.constants import SPAN_MEASURED_KEY
+from ddtrace.internal.utils import get_argument_value
+from ddtrace.pin import Pin
+from ddtrace.vendor.wrapt import wrap_function_wrapper as _w
+
+from .. import trace_utils
+from ...ext import SpanTypes
+
+
+config._add("graphql", dict(_default_service="graphql"))
+
+
+graphql_version_str = getattr(graphql, "__version__", "0.0.0")
+graphql_version = tuple([int(i) for i in graphql_version_str.split(".")])
+
+
+def patch():
+    if getattr(graphql, "_datadog_patch", False) or graphql_version < (2, 0):
+        return
+
+    setattr(graphql, "_datadog_patch", True)
+
+    graphql_module = "graphql.graphql"
+    graphql_func = "graphql_impl"
+    if graphql_version < (3, 0):
+        graphql_module = "graphql"
+        graphql_func = "graphql"
+
+    parse_execute_module = "graphql.graphql"
+    if (2, 1) <= graphql_version < (3, 0):
+        parse_execute_module = "graphql.backend.core"
+
+    validate_module = "graphql.validation"
+    if graphql_version < (2, 1):
+        validate_module = "graphql.graphql"
+    elif (2, 1) <= graphql_version < (3, 0):
+        validate_module = "graphql.backend.core"
+
+    resolve_module = "graphql.execution.execute"
+    # ExecutionContext.resolve_field was renamed to execute_field in graphql-core 3.2
+    resolve_func = "ExecutionContext.execute_field"
+    if graphql_version < (3, 0):
+        resolve_module = "graphql.execution.executor"
+        resolve_func = "resolve_field"
+    elif graphql_version < (3, 2):
+        resolve_func = "ExecutionContext.resolve_field"
```

Comment:
Can we only do a single version check and then duplicate the patch methods? This seems to be trying to be too clever and will be hard to change if one of the patching functions no longer works for a version.

Also it's hard to read the code and know immediately which functions are patched.